### ISSUE TITLE
Element size mismatch with TypeVarTuple and default arguments

### ISSUE THREAD

**Issue by cdce8p** [microsoft/pyright#3775]
**Describe the bug**
```
  error: Argument of type "*tuple[int]" cannot be assigned to parameter "args"
        of type "*Ts@run_callback" in function "run_callback"
    Type "*tuple[int]" cannot be assigned to type "*tuple[int, str]"
      Element size mismatch; expected 2 but received 1
```

**To Reproduce**
```python
from typing import Callable
from typing_extensions import Unpack, TypeVarTuple

Ts = TypeVarTuple("Ts")

def func(x: int, y: str = "Hello") -> None: ...

def run_callback(callback: Callable[[Unpack[Ts]], None], *args: Unpack[Ts]) -> None:
    # ...
    callback(*args)


run_callback(func, 3)  # error
```

**Expected behavior**
`run_callback(func, 3)` should be valid since `y` has a default argument.

**VS Code extension or command-line**
1.1.264

**Additional context**
For reference the relevant PEP section: https://peps.python.org/pep-0646/#type-variable-tuples-with-callable
However, AFAICT the specific interaction with default arguments in Callables isn't mentioned.

**Comment by erictraut** [comment]
Yeah, this won't work as you're hoping. Unlike `ParamSpec`, `TypeVarTuple` doesn't have any way to capture the notion of a parameter with a default. In your example, `Ts` either takes on the type of `tuple[int, str]` or `tuple[int]`; it can't be both. I suppose it would be theoretically possible to track both and have the constraint solver eliminate one or the other later in the process, but that would add significant complexity to something that's already very complex, so I'm not inclined to add that.

You can work around this by not using default arguments or using `@functools.partial` or a lambda wrapper to eliminate the parameters with default arguments.

**Comment by cdce8p** [comment]
> I suppose it would be theoretically possible to track both and have the constraint solver eliminate one or the other later in the process, but that would add significant complexity to something that's already very complex, so I'm not inclined to add that.

That's unfortunate, I suspect it will limit the usefulness of TypeVarTuples in Callables. In just one project, I already came across multiple errors because of it. Especially since it's such a common pattern with async code in particular. E.g. https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.call_later

Of course `functools.partial` would work, but that negates the purpose of adding type hints in the first place as I would lose type checking completely.

**Comment by erictraut** [comment]
The use of `functools.partial` shouldn't eliminate type checking. Pyright (and mypy) have special knowledge of `functools.partial`, so it should work fine.

**Comment by cdce8p** [comment]
> The use of `functools.partial` shouldn't eliminate type checking. Pyright (and mypy) have special knowledge of `functools.partial`, so it should work fine.

My experience so far has been to avoid partial as much as possible and unfortunately, it doesn't seem to work in this case
```py
from functools import partial
run_callback(partial(func, 3))
```

```
  test.py:16:1 - error: Argument of type "tuple[()]" cannot be assigned to parameter "args" of type "*Ts@run_callback" in function "run_callback"
    Type "tuple[()]" cannot be assigned to type "*tuple[str]"
      Element size mismatch; expected 1 but received 0 (reportGeneralTypeIssues)
```

**Comment by erictraut** [comment]
Ah, I forgot that `partial` has this weird behavior with default arguments. Yeah, that won't work. A lambda will work though.

```python
run_callback(lambda x: func(x), 3)
```

**Comment by cdce8p** [comment]
> A lambda will work though.

I'm not really convinced that is a good option. The arguments of lambda are untyped, so we again lose type checking for `*args`.
```py
reveal_type(lambda x: func(x))  # (x: Unknown) -> None
```

More importantly, using a lambda would complicate an already difficult to understand function call. Especially when passing more than one or two arguments. In the end, it's not a feasible solution IMO.

--

> Yeah, this won't work as you're hoping. Unlike `ParamSpec`, `TypeVarTuple` doesn't have any way to capture the notion of a parameter with a default. In your example, `Ts` either takes on the type of `tuple[int, str]` or `tuple[int]`; it can't be both. I suppose it would be theoretically possible to track both and have the constraint solver eliminate one or the other later in the process, but that would add significant complexity to something that's already very complex, so I'm not inclined to add that.

I don't know the internals of pyright. However, from my rough understanding of things I'm not sure it needs to be that complicated. Of course, it's necessary to track whether a parameter has a default argument. Once you do though, it's only necessary to compare the first x arguments and end the check. It should help that TypeVarTuples can only be used for `*args`, so the argument order will always match.

```py
def func(x: int, y: str = "Hello", z: float = 1) -> None: ...

run_callback(func, 3)  # ok
run_callback(func, 3, "World")  # ok
run_callback(func, 3, 2.4)  # error -> y missing
```

**Comment by cdce8p** [comment]
Thanks for the detailed answer! I was perhaps a bit too narrowly focused on my example without considering the general case. It's not a full nor particularly good solution, but what if support for default arguments is limited to cases where `Ts` isn't part of the return type? In these instances, it should only be necessary to keep track of defaults and do input type checking. That could allow supporting the (probably) most common case without adding ambiguity.

I've opened https://github.com/python/typing/issues/1231 to discuss how to handle this case further.

**Comment by erictraut** [comment]
> it should only be necessary to keep track of defaults

As I said above, TypeVarTuples don't capture parameter signatures, so it doesn't make sense to speak of "keeping track of defaults". TypeVarTuples capture tuples, and there is no concept of a "default" with a tuple.

You should think of this as a transform that converts a callable's parameter signature into a tuple. This transform fails for some signatures, such as those with keyword-only parameters or `**kwargs`. Signatures that include default arguments can be converted to multiple different tuples. Pyright's current implementation chooses the most natural conversion, which ignores the presence of the default arguments.
